pkg/clients/clients.go

```diff
@@ -217,35 +216,54 @@ func (dcs *DynamicClientSet) namespaced(gvk schema.GroupVersionKind) (bool, erro
 	return true, fmt.Errorf("failed to discover namespace info for %s", gvk)
 }
 
-func (dcs *DynamicClientSet) getServerResources(gvs ...schema.GroupVersion) ([]*v1.APIResourceList, error) {
-	var resources []*v1.APIResourceList
-	// TODO(lblackstone): Ensure that retry behavior is predictable.
-	err := retry.SleepingRetry(
-		func(i uint) (err error) {
-			if len(gvs) > 0 {
-				var resource *v1.APIResourceList
-				for _, gv := range gvs {
-					resource, err = dcs.DiscoveryClientCached.ServerResourcesForGroupVersion(gv.String())
-					if err != nil {
-						return
-					}
-
-					resources = append(resources, resource)
+func (dcs *DynamicClientSet) getServerResources(gvs ...schema.GroupVersion,
+) (resourceLists []*v1.APIResourceList, err error) {
```

Comment:
I think that I will shoot myself in the foot with a "return an `error` _and_ a valid return value" strategy here. What do you think of passing a callback that lets you process each entry, as `os.Walk` does? Then it could be up to the caller to keep track of errors.

Reply:
My thought was that the caller should be able to ignore GroupDiscoveryErrors, so it would only return an error in case of catastrophic failure. I can explicitly clear the resourceLists in those cases if you think that would be clearer.

Reply:
I am not worried about that case so much as _any_ case where we return an `error` _and_ a return value. I would personally greatly prefer the `os.Walk` approach, which AFAIK is "the" idiom to use in these situations. I 100% promise you that if we do what's currently shown here, I will mess this up if I have to add a callsite for this function. :)

Reply:
I took another pass and simplified the logic quite a bit. None of the functions return both valid data and an error at this point. LMK what you think.